src/main/java/com/redhat/devtools/lsp4ij/features/semanticTokens/viewProvider/LSPSemanticTokenPsiElement.java

```diff
@@ -0,0 +1,152 @@
+     *
+     * @param semanticToken the semantic token
+     */
+    LSPSemanticTokenPsiElement(@NotNull LSPSemanticToken semanticToken) {
+        super(semanticToken.getFile(), semanticToken.getTextRange());
+        this.isFile = Objects.equals(semanticToken.getFile().getTextRange(), semanticToken.getTextRange());
+        this.semanticToken = semanticToken;
+    }
+
+    @Override
+    public boolean isPhysical() {
+        // These do represent real text ranges in physical files
+        return true;
+    }
+
+    @Override
+    public boolean isValid() {
+        // Tie the validity of this element to that of its containing file
+        return semanticToken.getFile().isValid();
+    }
+
+    @Override
+    public boolean canNavigate() {
+        // References can be navigated
+        return semanticToken.getElementType() == LSPSemanticTokenElementType.REFERENCE;
+    }
+
+    @Override
+    @Nullable
+    public PsiElement getNameIdentifier() {
+        // If this is a declaration, return this element as the name identifier
+        return semanticToken.getElementType() == LSPSemanticTokenElementType.DECLARATION ? this : null;
+    }
+
+    @Override
+    @Nullable
+    public ASTNode getNode() {
+        // This is lazy-initialized because it needs the element text which is itself lazy-initialized
+        if (node == null) {
+            synchronized (this) {
+                if (node == null) {
+                    node = new LeafPsiElement(semanticToken.getElementType(), getText());
+                }
+            }
+        }
+
+        return node;
+    }
+
+    @Override
+    @NotNull
+    public String getText() {
+        // Optimization for full-file elements to avoid copying the full file text
+        if (isFile) return getContainingFile().getText();
+
+        // This is lazy-initialized because to avoid having to derive it until/unless needed
+        if (text == null) {
+            synchronized (this) {
+                if (text == null) {
+                    String workingText = "";
+                    TextRange textRange = getTextRange();
+                    if (textRange != null) {
+                        Document document = LSPIJUtils.getDocument(this);
+                        CharSequence documentChars = document != null ? document.getCharsSequence() : null;
+                        CharSequence textChars;
+                        if (documentChars != null) {
+                            int startOffset = Math.max(textRange.getStartOffset(), 0);
+                            int endOffset = Math.min(textRange.getEndOffset(), documentChars.length());
+                            textChars = documentChars.subSequence(startOffset, endOffset);
+                        } else {
+                            textChars = null;
+                        }
+                        if (textChars != null) {
+                            workingText = textChars.toString();
```

Comment:
And now we should only have to copy/cache the minimal text for exactly the semantic token text range.